### ISSUE TITLE
Make possible loading model with custom gradient

### ISSUE THREAD

**Issue by artem-s-shevchenko** [tensorflow/tensorflow#40166]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): tf2.1
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
Currently when a saved model is loaded, a custom gradient is not loaded.
Warnings like 
"WARNING: Importing a function _function-name_ with ops with custom gradients. Will likely fail if a gradient is requested." 
are shown and the model is failing during training with error 
"LookupError: No gradient defined for operation '_namespace_/IdentityN' (op type: IdentityN)"

Link to TODO in code:
https://github.com/tensorflow/tensorflow/blob/b794497e61fdb976448c562e9b87b09e24cbfabf/tensorflow/python/saved_model/function_deserialization.py#L351
It would be beneficial to have it working.

**Will this change the current api? How?**
No

**Who will benefit with this feature?*
People using custom gradient

**Any Other info.**

**Comment by lgeiger** [comment]
@Saduf2019 @rmothukuru @jvishnuvardhan is there any progress on this issue?

We are using custom gradients in many places in https://github.com/larq/larq to support training of quantized neural networks which currently prevents us from switching to `SavedModel` as a canonical serialization format for our models.

I am unfamiliar with this part of the codebase but I would be happy to send PRs to fix this if you have a rough idea of how this could be fixed.

**Comment by Lewington-pitsos** [comment]
Just a note, the implementation from: https://github.com/qubvel/efficientnet seems not to have this error, so switching to that implementation is a possible hacky workaround for now.

**Comment by asterisk37n** [comment]
That's nice. For some reason, its pretrained network shows 100x bigger loss than tensorflow's efficientnet.

**Comment by ardila** [comment]
+1 Isn't keras efficientnet one of the most commonly used models? Seems like a big hole in the SavedModel format

**Comment by zaccharieramzi** [comment]
Just stumbled upon this issue.
I am using a function which absolutely needs a custom gradient ([Non-uniform fourier transform](https://github.com/zaccharieramzi/tfkbnufft)), and I also absolutely need to save the model with the `SaveModel` format because I want to keep the optimizer's state (I am training in multiple sessions in an HPC setup, so I have to checkpoint).

I am considering using [this SO answer](https://stackoverflow.com/a/49504376/4332585) to manually save the optimizer weights. Has anyone tried this for tf2.x?

It would however be much easier with the `SaveModel` format out-of-the-box. Do you have any updates on this @k-w-w ?

**Comment by petrkoutnycz** [comment]
What's the status on this please? I've run several quite expensive computations and all of them ended up like this which is annoying.

**Comment by ihowell** [comment]
+1 It's making experiments quite hard as I now have to save all arguments to a file and reload them to rebuild the model from scratch, because I need the gradients from my decode function.

**Comment by frazane** [comment]
I'm having the same issue when I load a model that uses a Tensorflow Probability layer (a custom IndependentGamma layer, identical class to IndependentNormal but with Gamma distribution). 

`WARNING:tensorflow:Importing a function (__inference_random_gamma_3606) with ops with custom gradients. Will likely fail if a gradient is requested.`

**Comment by 4sfaloth** [comment]
Any news on this? I have exported an efficientnet from keras (.h5) to tensorflow (.pb) format for **inference only**.  It does work, but I get tons of those annoying "_with ops with custom gradients. Will likely fail if a gradient is requested._" warnings. If I'm not mistaken gradients are only required for training, so perhaps there's a way to load the model for inference model so they are ignored?

**Comment by maxcrous** [comment]
Any update on this @k-w-w? 
I am running into the same issue with tensorflow 2.4.1. 
@zaccharieramzi 's solution works nicely for saving optimizer state during training. 
But as @4sfaloth mentions, when using the saved_model format for storing inference models, model loading causes many unnecessary warnings.

**Comment by Luux** [comment]
This is really a major problem with the tf SavedModel format as you cannot even continue training a saved EfficientNet model currently.

Minimal example:

```
import tensorflow as tf
model = tf.keras.applications.EfficientNetB0()
model.compile()
model.save("broken")

model = tf.keras.models.load_model("broken")
```

**Comment by bhack** [comment]
It is experimental supported in nightly for the [next 2.6 release](https://www.tensorflow.org/guide/advanced_autodiff#custom_gradients)

`pip install tf-nightly` 

```python
import tensorflow as tf
model = tf.keras.applications.EfficientNetB0()
model.compile()
tf.keras.models.save_model(
    model, 'broken',
    options=tf.saved_model.SaveOptions(experimental_custom_gradients=True))

model = tf.keras.models.load_model("broken")
```

**Comment by sparshgarg23** [comment]
I was trying to run an example in tensorflow object detection api where in they were using a pretrained centernet model.Any ideas what could be the possible reason behind this.

**Comment by gadagashwini** [comment]
Issue is fixed in latest stable version Tf 2.8.2 
Please find the gist [here](https://colab.research.google.com/drive/1cwK36HOSFrBCUmlV4NGgTTTz9MsQ8WX3?usp=sharing). Thank you!
